core/src/test/java/org/elasticsearch/action/bulk/BulkProcessorRetryIT.java

```diff
@@ -0,0 +1,193 @@
+                .put("threadpool.search.queue_size", 1)
+                .put("threadpool.suggest.queue_size", 1)
+                .put("threadpool.percolate.queue_size", 1)
+                .put("threadpool.management.queue_size", 1)
+                .put("threadpool.refresh.queue_size", 1)
+                .put("threadpool.warmer.queue_size", 1)
+                .put("threadpool.snapshot.queue_size", 1)
+                .put("threadpool.force_merge.queue_size", 1)
+                .build();
+    }
+
+
+    public void testSimulatedBulkRejectionLoadWithoutBackoff() throws Throwable {
+        int numberOfAsyncOps = randomIntBetween(600, 700);
+        final CountDownLatch latch = new CountDownLatch(numberOfAsyncOps);
+        final Set<Object> responses = Collections.newSetFromMap(new ConcurrentHashMap<>());
+
+        BulkProcessor bulkProcessor = BulkProcessor.builder(client(), new BulkProcessor.Listener() {
+            @Override
+            public void beforeBulk(long executionId, BulkRequest request) {
+                // no op
+            }
+
+            @Override
+            public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+                responses.add(response);
+                latch.countDown();
+            }
+
+            @Override
+            public void afterBulk(long executionId, BulkRequest request, Throwable failure) {
+                responses.add(failure);
+                latch.countDown();
+            }
+        }).setBulkActions(1)
+                .setConcurrentRequests(100)
+                .build();
+        indexDocs(bulkProcessor, numberOfAsyncOps);
+        latch.await(10, TimeUnit.SECONDS);
+        bulkProcessor.close();
+
+        assertThat(responses.size(), equalTo(numberOfAsyncOps));
+
+        // validate all responses
+        for (Object response : responses) {
+            if (response instanceof BulkResponse) {
+                BulkResponse bulkResponse = (BulkResponse) response;
+                for (BulkItemResponse bulkItemResponse : bulkResponse.getItems()) {
+                    if (bulkItemResponse.isFailed()) {
+                        BulkItemResponse.Failure failure = bulkItemResponse.getFailure();
+                        Throwable rootCause = ExceptionsHelper.unwrapCause(failure.getCause());
+                        // we're overwhelming the pool, that's the exception we want to get
+                        if (rootCause instanceof EsRejectedExecutionException == false) {
+                            fail("Got unexpected reason [" + failure.getMessage() + "]");
+                        }
+                    }
+                }
+            } else {
+                Throwable t = (Throwable) response;
+                // we're not expecting any other errors
+                fail("Got unexpected failure: " + t);
+            }
+        }
+    }
+
+    public void testSimulatedBulkRejectionLoadWithBackoff() throws Throwable {
```

Comment:
So far as I can tell this doesn't simulate anything - it causes the rejections outright. So `testBulkRejectionLoadWithBackoff`?

Also, can these two test cases share any code? They are totally two sides of the same coin but I think we're less likely to introduce errors if you pull some of the common code into methods they can call. Lots of other Elastic reviewers won't comment on this but I've waded through lots of copy and pasted test code so I'm extra sensitive to it.